File: src/lib/lagrass.cpp
/***************************************************************************
 *   Copyright (C) 2007 by: Tim Sutton        tim@linfiniti.com            *
 *                          Jason Jorgenson   arkygeek@gmail.com           *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

#include "lagrass.h"
#include "lautils.h"
#include "la.h"
//#include "lagrassprocess.h"

#include <QString>
#include <QStringList>
#include <QSettings>
#include <QFile>
#include <QTextStream>
#include <QProcess>
#include <QStringList>
#include <QDebug>


LaGrass::LaGrass() : QObject()
{
}
LaGrass::~LaGrass()
{

}

QStringList LaGrass::getMapsetList()
{
  QString myCommand = "g.mapsets";
  QStringList myArguments;
  myArguments << "-l";
  //run the command ignoring error logs
  QString myResult = runCommand(myCommand,myArguments);
  myResult = myResult.simplified();
  QStringList myList = myResult.split(QRegExp("\\s+")); // splits on white space
  return myList;
}

QStringList LaGrass::getRasterList(QString theMapset, bool thePrependMapsetFlag /*=true */)
{
  QString myCommand = "g.list";
  QStringList myArguments;
  myArguments << "type=rast";
  myArguments << "mapset=" + theMapset;
  //run the command ignoring error logs
  QString myResult = runCommand(myCommand,myArguments);
  //check for no files
  if (myResult.contains("no raster files available"))
  {
    //return an empty list
    return QStringList();
  }
  //strip off grass decoration crap...
  myResult.replace("raster files available in mapset " + theMapset + ":","");
  myResult = myResult.simplified();
  QStringList myList = myResult.split(QRegExp("\\s+")); // splits on white space
  QStringList myFinalList;
  //prepend the mapset name to each raster name
  if (thePrependMapsetFlag)
  {
    QStringListIterator myIterator(myList);
    while (myIterator.hasNext())
    {
      QString myString = myIterator.next();
      myFinalList << myString + "@" + theMapset;
    }
  }
  return myFinalList;
}



bool LaGrass::copyMap(QString theOriginalRaster, QString theCopy)
{
  QString myCommand = "g.copy";
  QStringList myArguments;
  myArguments << "rast=" + theOriginalRaster + "," + theCopy;
  qDebug(myCommand.toLocal8Bit() + "  " + myArguments.join(" ").toLocal8Bit());
  QString myErrorLog;
  QString myResult = runCommand(myCommand,myArguments,myErrorLog);
  qDebug(myResult.toLocal8Bit());
  if (myErrorLog.isEmpty())
  {
    return true;
  }
  else
  {
    return false;
  }
}

bool LaGrass::createFrictionMap(QString theBaseRaster,QString theOutputRaster)
{
  //r.mapcalc "laFrictionMap = if(isnull(laDEM), null(), 1)"
  QString myCommand = "r.mapcalc";
  QStringList myArguments;
  myArguments << theOutputRaster + " = if(isnull(" + theBaseRaster + "), null(), 1)";
  qDebug(myCommand.toLocal8Bit() + "  " + myArguments.join(" ").toLocal8Bit());
  QString myErrorLog;
  QString myResult = runCommand(myCommand,myArguments,myErrorLog);
  qDebug(myResult.toLocal8Bit());
  if (myErrorLog.isEmpty())
  {
    return true;
  }
  else
  {
    return false;
  }
}

bool LaGrass::createInverseMask(float theMin, QString theMaskRaster)
{
  QString myCommand = "r.mapcalc";
  QStringList myArguments;
  myArguments << "laCostMapReclassed=if(laWalkCost>" + QString::number(theMin) + ",1,0)";
  qDebug(myCommand.toLocal8Bit() + "  " + myArguments.join(" ").toLocal8Bit());
  QString myErrorLog;
  QString myResult = runCommand(myCommand,myArguments,myErrorLog);
  qDebug(myResult.toLocal8Bit());

  createCombinedMask("laCostMapReclassed", theMaskRaster);

  if (myErrorLog.isEmpty())
  {
    return true;
  }
  else
  {
    return false;
  }
}

bool LaGrass::createMask(QString theCostSurface, QString theMaskRaster)
{
  //r.mapcalc "laFrictionMap = if(isnull(laDEM), null(), 1)"
  QString myCommand = "r.mapcalc";
  QStringList myArguments;
  QString myMaskName = "tmpMask";
  myArguments << myMaskName + "=" + theCostSurface + "*" + theMaskRaster;
  qDebug(myCommand.toLocal8Bit() + "  " + myArguments.join(" ").toLocal8Bit());
  QString myErrorLog;
  QString myResult = runCommand(myCommand,myArguments,myErrorLog);
  qDebug(myResult.toLocal8Bit());
  if (myErrorLog.isEmpty())
  {
    return true;
  }
  else
  {
    return false;
  }
}

bool LaGrass::createCombinedMask(QString theCostSurface, QString theMaskRaster)
{
  QString myCommand = "r.mapcalc";
  QStringList myArguments;
  QString myMaskName = "laLeftOver";
  myArguments << myMaskName + "=" + theCostSurface + "*" + theMaskRaster;
  qDebug(myCommand.toLocal8Bit() + "  " + myArguments.join(" ").toLocal8Bit());
  QString myErrorLog;
  QString myResult = runCommand(myCommand,myArguments,myErrorLog);
  qDebug(myResult.toLocal8Bit());
  if (myErrorLog.isEmpty())
  {
    return true;
  }
  else
  {
    return false;
  }
}

bool LaGrass::mergeMaps(QString theLeftoversGoHere)
{


  QString myErrorLog;
  QString myCommand = "r.mapcalc";

  // first we need to make the remaining land available
  // by changing all the null values to zero, and then
  // adding the two together.  (null + 1 is still null)

  QStringList myArguments;
  myArguments << "mergeTmp = if( isnull ( " + theLeftoversGoHere  + "),0," + theLeftoversGoHere + ")";
  QString myResult = runCommand(myCommand,myArguments,myErrorLog);
  qDebug(myResult.toLocal8Bit());

  myArguments.clear();
 //+ if( isnull ( laLeftOver),0,laLeftOver)\'";
  myArguments << "laLeftOverTmp = if( isnull ( laLeftOver),0,laLeftOver)";
  myResult = runCommand(myCommand,myArguments,myErrorLog);
  qDebug(myResult.toLocal8Bit());

  myArguments.clear();
  myArguments << "laCombinedMasks = laLeftOverTmp + mergeTmp";
  myResult = runCommand(myCommand,myArguments,myErrorLog);
  qDebug(myResult.toLocal8Bit());

  // clean up temp files
  removeFile("mergeTmp");
  removeFile("laLeftoverTmp");

  if (myErrorLog.isEmpty())
  {
    return true;
  }
  else
  {
    return false;
  }
}

float LaGrass::getArea(QString theLayerName)
{
  logMessage("method ==> void LaGrass::getArea(float theArea)");
  //r.stats -a -n fs=,
  QString myCommand = "r.stats";
  QStringList myArguments;
  myArguments << "-a" << "-n" << "fs=," << "input="+theLayerName;
  qDebug(myCommand.toLocal8Bit() + "  " + myArguments.join(" ").toLocal8Bit());
  QString myErrorLog;
  QString myResult = runCommand(myCommand,myArguments,myErrorLog);
  qDebug(myResult.toLocal8Bit());
  if (myErrorLog.isEmpty())
  {
    return 0;
  }
  else
  {
    qDebug() << "myResult: " << myResult;
    myResult = myResult.simplified();
    qDebug() << "myResult.simplified(): " << myResult;
    //put each line of output into a list entry
    QStringList myList = myResult.split(QRegExp("\\s+"));
    qDebug() << "myList: " << myList;
   // if (myList.count() < 1)
    //{
      //row should be like 3,32323 (class,area)
      QStringList myList2 = myList.at(1).split((","));
      qDebug() << "myList2: " << myList2;
      //if (myList2.count() < 1)
      //{
        //get only area
        float myArea = myList2.at(1).toFloat() * 0.0001; // adjust for hectares
        qDebug() << "Area (returnValue): " << myArea;
        return myArea;
      //}
    //}
  }
  return 0;
}

bool LaGrass::makeWalkCost(int theX, int theY, QString theDEM)
{
  //logMessage("method ==> void LaGrass::makeWalkCost(int theX, int theY)");
  //r.walk max_cost=20000 elevation=dem_patched_filled@PERMANENT friction=theFrictionMap output=rwalkResultsSlopeMax20kFMap coordinate=744800,3611100 percent_memory=100 nseg=4 walk_coeff=0.72,6.0,1.9998,-1.9998 lambda=0 slope_factor=-0.2125 -k
  //qDebug() << "Coordinates: " << theX << "," << theY;
  QString myElevationMap=theDEM;
  QString myFrictionMap=theDEM;

  QString myCommand = "r.walk";
  QStringList myArguments;
  myArguments << "max_cost=18000"
              << "elevation=" + myElevationMap
              << "friction=" + myFrictionMap
              << "output=laWalkCost"
              << "coordinate=" + QString::number(theX) + "," + QString::number(theY)
              << "percent_memory=100"
              << "nseg=4"
              << "walk_coeff=0.72,6.0,1.9998,-1.999"
              << "lambda=0"
              << "slope_factor=-0.2125"
              << "-k"
              << "--overwrite";
  qDebug(myCommand.toLocal8Bit() + "  " + myArguments.join(" ").toLocal8Bit());

  QString myErrorLog;
  QString myResult = runCommand(myCommand,myArguments,myErrorLog);
  //qDebug(myResult.toLocal8Bit());
  //LaGrassProcess myLaGrassProcess;
  if (myErrorLog.isEmpty())
  {
    return true;
  }
  else
  {
    return false;
  }
}

void LaGrass::makeEuclideanCost(int theX, int theY)
{
  logMessage("method ==> void LaGrass::makeEuclideanCost(int theX, int theY)");
}

void LaGrass::makePathDistanceCost(int theX, int theY)
{
  logMessage("method ==> void LaGrass::makePathDistanceCost(int theX, int theY)");
}

void LaGrass::writeMetaData(QStringList theMetaData)
{
  logMessage("method ==> void LaGrass::writeMetaData(QString theValue)");
  /*  ---> BASH example <---
    # write info to history section of raster map
    r.support map=catchment history="Radius was ${radius}"
    r.support map=catchment history="This results in ${area} sq meters and the target was ${land_reqrd_total}"

   */

  // iterate through the list, adding each to the history.
  //QString myCommand = "r.support";
  //QStringList myArguments;
  //QString myMaskName = "laLeftOver";
  //myArguments << myMaskName + "=" + theCostSurface + "*" + theMaskRaster;
  //qDebug(myCommand.toLocal8Bit() + "  " + myArguments.join(" ").toLocal8Bit());
  //QString myErrorLog;
  //QString myResult = runCommand(myCommand,myArguments,myErrorLog);
  //qDebug(myResult.toLocal8Bit());
  //if (myErrorLog.isEmpty())
  //{
    //return true;
  //}
  //else
  //{
  //  return false;
  //}
}

bool LaGrass::reclass(QString theRaster, int theMax)
{
  qDebug() << "reclass invoked";
  //r.mapcalc "laFrictionMap = if(isnull(laDEM), null(), 1)"
  QString myCommand = "r.mapcalc";
  QStringList myArguments;
  myArguments << "laCostMapReclassed=if(" + theRaster + "<" + QString::number(theMax) + ",1,0)";
  qDebug(myCommand.toLocal8Bit() + "  " + myArguments.join(" ").toLocal8Bit());
  QString myErrorLog;
  QString myResult = runCommand(myCommand,myArguments,myErrorLog);
  qDebug(myResult.toLocal8Bit());
  if (myErrorLog.isEmpty())
  {
    return true;
  }
  else
  {
    return false;
  }
}

bool LaGrass::removeFile(QString theFile)
{
  QString myCommand = "g.remove";
  QStringList myArguments;
  myArguments << "rast=" + theFile;
  qDebug(myCommand.toLocal8Bit() + " " + myArguments.join(" ").toLocal8Bit());
  QString myErrorLog;
  QString myResult = runCommand(myCommand,myArguments,myErrorLog);
  qDebug(myResult.toLocal8Bit());
  if (myErrorLog.isEmpty())
  {
    return true;
  }
  else
  {
    return false;
  }
}

//
// Utility functions
//

void LaGrass::logMessage(QString theMessage)
{
  emit message(theMessage);
}

QString LaGrass::runCommand(QString theCommand,
    QStringList theArguments,
    QString &theErrorLog /*=""*/)
{
  #ifdef Q_OS_MACX
  QString myProgram = "/Applications/GRASS.app/Contents/Resources/bin/" + theCommand;
  #else
  QString myProgram = "/usr/lib/grass/bin/" + theCommand;
  #endif
  //windows users can wallow in self pity for now...

  QProcess myProcess;
  myProcess.start(myProgram, theArguments);

  if (!myProcess.waitForStarted())
  {
    logMessage("The process never started.....aaargh");
  }

  while (myProcess.waitForReadyRead(-1))
  {
  }

  QString myLog;
  //log
  myProcess.setReadChannel(QProcess::StandardOutput);
  QByteArray myArray = myProcess.readAll();
  myLog.append(myArray);
  //errors
  myProcess.setReadChannel(QProcess::StandardError);
  myArray = myProcess.readAll();
  theErrorLog.append(myArray);
  // get rid of some grass decorations...
  qDebug() <<"GRASS> " << theCommand << theArguments;
  qDebug() <<"Results: " << myLog;
  myLog.replace("----------------------------------------------","");

  return myLog;
}
QString LaGrass::runCommand(QString theCommand,
    QStringList theArguments)
{
  QString myErrors;
  QString myLog = runCommand(theCommand, theArguments, myErrors);
  return myLog;
}

// reclass

File: landuse_analyst/lib/lagrass.py
from qgis.PyQt.QtCore import QObject, pyqtSlot, QProcess, QStringListModel, QRegExp
from typing import Dict, List
import os

class LaGrass(QObject):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.diets = {}

    @pyqtSlot(str, result=object)
    def getDiet(self, guid: str):
        return self.diets.get(guid)

    @pyqtSlot(str, object)
    def addDiet(self, guid: str, diet):
        self.diets[guid] = diet

    @pyqtSlot(str)
    def removeDiet(self, guid: str):
        if guid in self.diets:
            del self.diets[guid]

    @pyqtSlot(result=dict)
    def getDiets(self) -> Dict[str, object]:
        return self.diets

    def runCommand(self, command: str, arguments: List[str], error_log: str = "") -> str:
        program = f"/usr/lib/grass/bin/{command}"
        process = QProcess()
        process.start(program, arguments)

        if not process.waitForStarted():
            self.logMessage("The process never started.....aaargh")
            return ""

        while process.waitForReadyRead(-1):
            pass

        log = ""
        process.setReadChannel(QProcess.StandardOutput)
        log += process.readAll().data().decode()
        process.setReadChannel(QProcess.StandardError)
        error_log += process.readAll().data().decode()

        log = log.replace("----------------------------------------------", "")
        return log

    def getMapsetList(self) -> List[str]:
        command = "g.mapsets"
        arguments = ["-l"]
        result = self.runCommand(command, arguments)
        result = result.strip()
        return result.split()

    def getRasterList(self, mapset: str, prepend_mapset_flag: bool = True) -> List[str]:
        command = "g.list"
        arguments = ["type=rast", f"mapset={mapset}"]
        result = self.runCommand(command, arguments)

        if "no raster files available" in result:
            return []

        result = result.replace(f"raster files available in mapset {mapset}:", "").strip()
        raster_list = result.split()
        if prepend_mapset_flag:
            return [f"{raster}@{mapset}" for raster in raster_list]
        return raster_list

    def copyMap(self, original_raster: str, copy: str) -> bool:
        command = "g.copy"
        arguments = [f"rast={original_raster},{copy}"]
        error_log = ""
        result = self.runCommand(command, arguments, error_log)
        return error_log == ""

    def createFrictionMap(self, base_raster: str, output_raster: str) -> bool:
        command = "r.mapcalc"
        arguments = [f"{output_raster} = if(isnull({base_raster}), null(), 1)"]
        error_log = ""
        result = self.runCommand(command, arguments, error_log)
        return error_log == ""

    def createInverseMask(self, min_value: float, mask_raster: str) -> bool:
        command = "r.mapcalc"
        arguments = [f"laCostMapReclassed=if(laWalkCost>{min_value},1,0)"]
        error_log = ""
        result = self.runCommand(command, arguments, error_log)
        self.createCombinedMask("laCostMapReclassed", mask_raster)
        return error_log == ""

    def createMask(self, cost_surface: str, mask_raster: str) -> bool:
        command = "r.mapcalc"
        mask_name = "tmpMask"
        arguments = [f"{mask_name}={cost_surface}*{mask_raster}"]
        error_log = ""
        result = self.runCommand(command, arguments, error_log)
        return error_log == ""

    def createCombinedMask(self, cost_surface: str, mask_raster: str) -> bool:
        command = "r.mapcalc"
        mask_name = "laLeftOver"
        arguments = [f"{mask_name}={cost_surface}*{mask_raster}"]
        error_log = ""
        result = self.runCommand(command, arguments, error_log)
        return error_log == ""

    def mergeMaps(self, leftovers_go_here: str) -> bool:
        error_log = ""
        command = "r.mapcalc"

        arguments = [f"mergeTmp = if(isnull({leftovers_go_here}),0,{leftovers_go_here})"]
        result = self.runCommand(command, arguments, error_log)

        arguments = [f"laLeftOverTmp = if(isnull(laLeftOver),0,laLeftOver)"]
        result = self.runCommand(command, arguments, error_log)

        arguments = [f"laCombinedMasks = laLeftOverTmp + mergeTmp"]
        result = self.runCommand(command, arguments, error_log)

        self.removeFile("mergeTmp")
        self.removeFile("laLeftOverTmp")

        return error_log == ""

    def getArea(self, layer_name: str) -> float:
        self.logMessage("method ==> void LaGrass::getArea(float theArea)")
        command = "r.stats"
        arguments = ["-a", "-n", "fs=,", f"input={layer_name}"]
        error_log = ""
        result = self.runCommand(command, arguments, error_log)

        if error_log:
            return 0

        result = result.strip()
        result_list = result.split()
        if len(result_list) < 2:
            return 0

        area_list = result_list[1].split(",")
        if len(area_list) < 2:
            return 0

        area = float(area_list[1]) * 0.0001  # adjust for hectares
        return area

    def makeWalkCost(self, x: int, y: int, dem: str) -> bool:
        self.logMessage("method ==> void LaGrass::makeWalkCost(int theX, int theY)")
        command = "r.walk"
        arguments = [
            "max_cost=40000",
            f"elevation={dem}",
            "friction=frictionMap",
            "output=laWalkCost",
            f"coordinate={x},{y}",
            "percent_memory=100",
            "nseg=4",
            "walk_coeff=0.72,6.0,1.9998,-1.999",
            "lambda=0.75",
            "slope_factor=-0.2125",
            "-k",
            "--overwrite"
        ]
        error_log = ""
        result = self.runCommand(command, arguments, error_log)
        return error_log == ""

    def makeEuclideanCost(self, x: int, y: int):
        self.logMessage("method ==> void LaGrass::makeEuclideanCost(int theX, int theY)")

    def makePathDistanceCost(self, x: int, y: int):
        self.logMessage("method ==> void LaGrass::makePathDistanceCost(int theX, int theY)")

    def writeMetaData(self, meta_data: List[str]):
        self.logMessage("method ==> void LaGrass::writeMetaData(QString theValue)")

    def reclass(self, raster: str, max_value: int) -> bool:
        command = "r.mapcalc"
        arguments = [f"laCostMapReclassed=if({raster}<{max_value},1,0)"]
        error_log = ""
        result = self.runCommand(command, arguments, error_log)
        return error_log == ""

    def removeFile(self, file: str) -> bool:
        command = "g.remove"
        arguments = [f"rast={file}"]
        error_log = ""
        result = self.runCommand(command, arguments, error_log)
        return error_log == ""

    def logMessage(self, message: str):
        self.message.emit(message)